src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,188 @@
+            # Generate environment name if it wasn't not explicitly supplied,
+            #  and isn't disabled
+            env_name = self.generate_env_name()
+        if env_name not in os.environ:
+            return False, None
+        env_var = os.environ.get(env_name, None)
+        loaded_var: _T | None | str = env_var
+        if env_var and self.type is not None:
+            loaded_var = self.type(env_var)
+        if isinstance(loaded_var, (Table, TOMLDocument)):
+            # If we got a TOML table we probably want it in dictionary form
+            return True, loaded_var.value
+        return True, loaded_var
+
+    def _get_config(self):
+        e = self._root_parser._conf
+        for k in self._nest_path[1:]:
+            e = e[k]
+        if isinstance(e, (Table, TOMLDocument)):
+            # If we got a TOML table we probably want it in dictionary form
+            return e.value
+        return e
+
+
+class ConfigParser:
+    def __init__(
+        self,
+        *,
+        name: str,
+        file_path: Path | None = None,
+    ):
+        self.name = name
+        self.file_path = file_path
+        # Objects holding subparsers and options
+        self._options: dict[str, ConfigOption] = dict()
+        self._sub_parsers: dict[str, ConfigParser] = dict()
+        # Dictionary to cache read in config file
+        self._conf: TOMLDocument | None = None
+        # Information necessary to be able to nest elements
+        #  and add options in O(1)
+        self._root_parser: ConfigParser = self
+        self._nest_path = [name]
+
+    def read_config(
+        self,
+    ) -> None:
+        """Read and parse config file."""
+        if self.file_path is None:
+            raise ConfigParserError(
+                "ConfigParser is trying to read config file, but it doesn't have one"
+            )
+        LOGGER.debug(f"reading configuration file from {str(self.file_path)}")
+        try:
+            self._conf = tomlkit.parse(self.file_path.read_text())
+        except Exception as e:
+            raise ConfigSourceError(
+                f"An unknown error happened while loading '{str(self.file_path)}"
+                f"', please see the error: {e}"
+            )
+
+    @wraps(ConfigOption.__init__)
+    def add_option(
+        self,
+        *args,
+        **kwargs,
+    ) -> None:
+        kwargs["_root_parser"] = self._root_parser
+        kwargs["_nest_path"] = self._nest_path
+        new_option = ConfigOption(
+            *args,
+            **kwargs,
+        )
+        self._check_child_conflict(new_option.name)
+        self._options[new_option.name] = new_option
+
+    def _check_child_conflict(self, name: str) -> None:
+        if name in (self._options.keys() | self._sub_parsers.keys()):
+            raise ConfigParserError(
+                f"'{name}' subparser, or option conflicts with a child element of '{self.name}'"
+            )
+
+    def add_subparser(self, other: ConfigParser) -> None:
```

Comment:
what's the usage of subparser and do we expect users to add subparser in runtime?

from the tests I saw this function is always called along with the constructor. I'm thinking whether we just make `subparser` part of the `__init__` method and remove this `add` method.

Reply:
Yes, these need to updated at run-time. Imagine that SnowPark wants to add its own option to the Snowflake config file. That can only be done at run-time and only by SnowPark when it gets imported.
So the idea was that SnowPark can create its own root parser and then nest that under the Connector's root parser with probably the name `snowpark`.
So something along the lines of

```python
snowpark_cp = ConfigParser(name="snowpark")
snowpark_cp.add_option(name="use_async_queries", _type=lambda s: s.lower() == "true", choices=(True, False))
from snowflake.connector.config_parser import CONFIG_PARSER
CONFIG_PARSER.add_subparser(snowpark_cp)
```

So hopefully every library using the Python connector can just maintain their own parser and once they added all option just register it to the connector (they can possibly nest more of them to create separation of options).